### ISSUE TITLE
is it safe to use simplewall + windows firewall?

### ISSUE THREAD

**Issue by 0lm** [henrypp/simplewall#254]
hi there
is it safe to use simplewall and windows firewall in compination? i jsut feel more savf with that. it wont cause any problems, will it?

**Comment by beerisgood** [comment]
I didn't disable the Windows firewall even if simplewall say that is recommend and i don't have problems since long usage

**Comment by ltGuillaume** [comment]
How will you keep tabs on what actually gets through if you still have Windows Firewall enabled? Those rules may have priority over anything you set up in simplewall.

**Comment by beerisgood** [comment]
If a program try a connection, simplewall ask me.
And if i say no, it works.

So i don't see a problem here

**Comment by pwn0r** [comment]
If simplewall works then WF is not needed.  SimpleWall filtering should take priority over WF.

**Comment by beerisgood** [comment]
Thats right but you get a warning from Windows if the firewall is disabled.
Of course you can disable that warning too, but i don't a reason for that.

**Comment by ltGuillaume** [comment]
I've read that it's the other way around: Windows Firewall rules could take
priority over simplewall's, which is definitely not what you want.

**Comment by beerisgood** [comment]
Did you read my post? Can't confirm this

**Comment by ltGuillaume** [comment]
If you use it just to block third party applications, I guess leaving
Windows Firewall on should be alright. If the application adds its own
firewall rules (such as uTorrent), I'm not sure if simplewall will help. If
you use it to block Windows services, or use the blocklist in simplewall,
I'd be sure to disable Windows Firewall. I'm following the logic of what
I've read here and in the Wiki, but I haven't tested these use cases
myself.

**Comment by ltGuillaume** [comment]
@henrypp added this issue to the FAQ, but hasn't responded to it himself. So what's the verdict?

**Comment by TontyTon** [comment]
Hey everyone I had used Simplewall (SW) in parallel with Windows Firewall (WA). And I kept a close watch on Resource Monitor, for quite some time few months back, to observe the priorities.
I observed that the two firewall works independently.

For example-
If you had blocked some file in WF, then whatever you do in SW, file will not be able to access internet. And vice-versa.

They are acting as two 'walls', for a thing to pass and access outside world, it has to pass through both of the 'walls'.

**Comment by TontyTon** [comment]
Hey everyone I had used Simplewall (SW) in parallel with Windows Firewall (WA). And I kept a close watch on Resource Monitor, for quite some time few months back, to observe the priorities.
I observed that the two firewall works independently.

For example-
If you had blocked some file in WF, then whatever you do in SW, file will not be able to access internet. And vice-versa.

They are acting as two 'walls', for a thing to pass and access outside world, it has to pass through both of the 'walls'.

**Comment by ltGuillaume** [comment]
That's all good and well, but apparently there is some sort of priority system in the form of weights. So unless the rules by Windows Defender Firewall always have a lower weight than those by simplewall, you can never really know if the simplewall rules are actually honored.

So @henrypp (how) are you taking this into account?

**Comment by TontyTon** [comment]
@ltGuillaume Hey! I didn't understand what you are saying, I am just a basic user and many terms you used were read by me very first time.
What I said is based on experiment. I tried many combinations with system and third party apps (I thought might be MS had made some difference) and inferenced "They are acting as two 'walls', for a thing to pass and access outside world, it has to pass through both of the walls." from the observations.

**Comment by ltGuillaume** [comment]
I know, and I appreciate the effort you made. Essentially you're right, to a point, **but** there can be exceptions, as can be read in the [documentation](https://docs.microsoft.com/en-us/windows/desktop/fwp/filter-arbitration). This means that you can't be sure everything you don't want to get through is in fact blocked (by simplewall), unless proper precautions are already taken by @henrypp .

**Comment by TontyTon** [comment]
@ltGuillaume This is open source project. Everyone who know about the programming language should contribute to it. @henrypp had done a lot of work and importantly he is regularly updating the program to fix bugs, update UI, etc. And if there is any problem in the program and he know how to correct it, I think he will definitely do the problem solving.

And for the people who don't know the language, but understands the WFP system, should tell him how everything should work but keeping it open to suggestions.

@henrypp I think, you should think of adding an article about how simplewall is intended to work (programmatically) for 'the people who don't know the language'.

**Comment by ltGuillaume** [comment]
@TontyTon I think we agree on that (just to be clear: I have virtually no C++ experience, so I couldn't contribute with pull requests even if I wanted to). But there really is something odd going on here: we started out asking a question, after which @henrypp even added this to his FAQ! But he never properly answered it (or even answered at all!). All I'm trying to do is to be as specific as possible when it comes to posing the question, that's all I can do really. No I'm hoping that @henrypp could be equally precise in his answer. We *are* asking here whether simplewall actually does what it's created for, after all...

**Comment by TontyTon** [comment]
Maybe he himself doesn't know what happens when WF and SW contradicts, because he knows about his program not the exact working of the Windows. Thats why **he had recommended to turn WF off when using SW**.

>  We are asking here whether simplewall actually does what it's created for, after all...

Hey, **do justice**, SW is completely working, it is doing what it is made for. It is not necessary to keep WF onn (if someone{_myself included_} is suspicious about it).

**Comment by ltGuillaume** [comment]
That's too easy. The only thing I'm saying is that stuff like this creates confusion, which is awful for security software, thus it needs to be figured out properly and so documented.

Don't act like I'm attacking @henrypp , or discrediting simplewall, I'm not.

**Comment by henrypp** [comment]
>Maybe he himself doesn't know what happens when WF and SW

Maybe not, and i am try to explain this right now.

When WF enabled, his filters applied _before any_ WFP provider (SW as example).
Any kind of software can be changed WF filter configuration, not even viruses or any kind of harmfull software, it can be done by any software you are installed (steam, utorrent, [even nsis installer have feature like this](https://nsis.sourceforge.io/NsisFirewall_plug-in), etc).

So i recommend to disable WF, because i am not responsible for effects like this.

**Comment by tokariu** [comment]
when WF and SW are running parallel, and WF filters are applied before SW filters, it would be most logical if the last applied filters count. in that case SW filters are applied after WF, so SW filters overwrite WF filters.

in other software firewalls it is internally done like that: when it is about global or application rules, the most restrictive rules counts.

But the situation between two firewalls parallel (WF / SW) might be a little different and I think the last applied filters overwrite the first applied filters. and because windows firewall filters are always applied first, it's simplewalls filters that count.

Therefor, even when other software (steam, games, etc) apply their own rules in windows firewall, simplewall will overrule that wf-rule.

That's just my thoughts. Correct me if I'm wrong.

**Comment by ltGuillaume** [comment]
@tokariu @henrypp You're making a whole lot of assumptions here that I can't find evidence for, namely:
- That WF filters will be "applied" (as you say) _before_ Simplewall's. Who says? In [this piece of documentation](https://docs.microsoft.com/en-us/windows/desktop/fwp/filter-arbitration) it states that traffic traverses all sub-layers "from the highest priority to the lowest priority". Only if Simplewall has a lower priority than WF (and I haven't a clue as to whether that is the case), then that would mean that WF's filters would be "applied" _first_.
- That the order in which the filters are evaluated actually matters. The process is to 
> Evaluate all sub-layers even if a higher priority sub-layer has decided to block the traffic.
So, traffic traverses all sub-layers (of which SW is one), if SW blocks the traffic, it will _still_ be processed by WF! Or if indeed Simplewall has a lower priority than WF, still, both sub-layers will be traversed. That is because "block this traffic and that's the end of it" is not how the Windows Filtering Platform works in all cases: every single filter (rule) can set the flag `FWPS_RIGHT_ACTION_WRITE` to "whether a lower priority sub-layer is allowed to override the action".

Example (**assuming** WF has a _higher_ priority):
Simplewall has a rule to block some traffic. WF has a rule to permit that traffic and has **not** set the `FWPS_RIGHT_ACTION_WRITE` flag. Now, _IF_ WF's sub-layer does indeed have a higher priority than Simplewall's sub-layer, then the package **will be permitted** (even though it looks like it's blocked by Simplewall).

**Comment by tokariu** [comment]
@ltGuillaume indeed, a lot of assumptions. That's something that hopefully @henrypp can elaborate for us.

Otherwise, we just have to test and find out what happens in all those situations and derive some rules from it.

**Comment by cloudgreen** [comment]
I'm not sure I understand fully...

So if I select "Disable Windows Firewall" when enabling the filtering in Simplewall, does that mean that all the Inbound & Outbound rules defined in Windows Firewall gets ignored (disabled)?

Any insights appreciated!

**Comment by ltGuillaume** [comment]
That's the idea, yes.

**Comment by cloudgreen** [comment]
> That's the idea, yes.

Thanks a lot for super-fast response!  :)

A few more:

1. With Windows Firewall disabled, all rules have to be edited from within the Simplewall GUI, both inbound and outbound rules?
2.  Are all inbound connections also blocked by default, when using Whitelist mode? (I assume yes, since there is a setting for _"Allow inbound connections for all"_...)
3. I think the main Simplewall GUI should indicate whether apps are allowed IN or OUT... as it is now it only says "Allowed apps". Is this refering to outbound connections only?

**Comment by ltGuillaume** [comment]
1. Yes
2. Everything should be blocked, except the inbound rules enabled in Rules -> General configuration, and for the system rules you've enabled.
3. I'm not sure how this works, to be honest. Whereas Windows Firewall asks for permission to allow inbound connections (and not for outbound), it looks like the option "Allow listen connections" (enabled by default) actually makes simplewall do the exact opposite, i.e. always allow incoming, while preventing outgoing. I haven't thought much about this before, but you're right, it needs to be cleared up.

**Comment by ltGuillaume** [comment]
@henrypp @cloudgreen Your third point was a really really good one: I have no clue whatsoever how to get ANY feedback as whether I have allowed an application to listen for incoming connections as soon as I disable "Allow listen connections".

**Comment by cloudgreen** [comment]
@ltGuillaume Thanks, and thanks for your replies! :)

Yes, I'm (obviously) pretty new to Simplewall, and while I think it's a very nice piece of software that really fills a gap in the Windows Firewall feature set with the outgoing application notifications, some aspects are a bit opaque to newcomers. How does it relate to Windows Firewall and how does it replace all its features when you turn off Windows Firewall (as per the recommendation)? You need to be pretty technical or do a lot of research to understand the concepts in this thread with Windows Filtering Platform, API:s, sublayers etc... :)

The division between ingoing & outgoing rules are easily understood in Windows Firewall, but less so in Simplewall. Then again I believe Simplewall's entire "raison d'être" is the outgoing notification/blocking part, so maybe that's where the focus has been and still is (?).

Running Simplewall with Windows Firewall disabled, it would really help if incoming and outgoing rules could be more clearly created and represented in the GUI. It would be like Windows Firewall, but simplified in a very (simple? :) ) and powerful way.

Keep up the great work!

**Comment by cloudgreen** [comment]
...and just another wild thought: could the Windows Firewall GUI be used as a front-end for viewing and managing all the rules created by Simplewall? That would be a great combo with the best of both worlds: Simplewall for notifications and for easily creating new rules, Windows Firewall for managing and viewing incoming and outgoing rules on a very detailed technical level.

Or is that maybe not doable because of some technical WFP/sublayer problems? :)

**Comment by ltGuillaume** [comment]
Although I do not think that the "Firewall with Advanced Security" GUI can be used to view and/or manage other sublayers, it would indeed be great to have a "flat view" to see how the settings in simplewall translate to the actual WFP rules. @henrypp

**Comment by TontyTon** [comment]
@cloudgreen @ltGuillaume 

Allowing listen connections don't allow incoming connections. "Allow listen connections for all" checkmark in settings is just to avoid errors in third party apps (as stated in tooltip).

"Allow inbound connections for all" is the only way to allow all incoming connections and it is not checkmarked by default.
So, SW neither allow incoming connections nor outgoing connections by default (when 'Whitelist' is selected).

Refer https://stackoverflow.com/a/34073929 [ 'listen()' is listen connction and accept() is for inbound connection]

**Comment by ltGuillaume** [comment]
@TontyTon Thanks for clearing that up. I should have researched this more instead of causing confusion here, I apologize. I quickly tested it and drew the wrong conclusions.

**Comment by cloudgreen** [comment]
@TontyTon Ah, to be honest I thought that @ltGuillaume wrote about the same setting as I did, I didn't notice that he referred to another checkbox. :)

**Comment by motiontwelve** [comment]
I am happy to leave Windows Firewall disabled as long as that stupid Windows Security's "Action Needed" red symbol (shown below) would disappear.  Has anyone figured out how?
![image](https://user-images.githubusercontent.com/9596320/58400657-f9cc5880-8018-11e9-9aa9-9bedd00ad42f.png)

**Comment by perdolka** [comment]
@motiontwelve
You can hide this icon:
![зображення](https://user-images.githubusercontent.com/24720263/58408652-e223bd00-8076-11e9-8edd-7b0579bf8a47.png)

Register Simplewall as firewall (in Windows 10) is impossible:
https://github.com/henrypp/simplewall/issues/364

**Comment by motiontwelve** [comment]
@perdolka 
But is there another way to inform the user that Windows Defender is still running?

**Comment by GintokiHub** [comment]
with the search engine of duckduckgo I found various solutions to shutting up the windows 'security alert' notifications. 

What I did if I remember correctly was simply right clicking that icon and clicking disable 'notificatons'
though I am not sure if that was how it went.

However here was the query I used.
: "majorgeeks disable notification windows firewall is disabled"


Also Majorgeeks probably has a couple different apps, which do this for you.

**Comment by webinger** [comment]
when i try to install 3.1.2, windows defender won't let me. on virustotal.com 3 engines claim to find a virus:  
Bkav: HW32.Packed.
Cylance: Unsafe
Microsoft: Trojan:Win32/Wacatac.C!ml
https://www.virustotal.com/gui/file/33a7340c0da32d688655fb6f1c008c2f3d0d8f7a82011f16fe7f01d0a2082f5a/detection

I installed it on another machine some weeks ago without any problems. any ideas?

**Comment by ltGuillaume** [comment]
False positive. Just download the zip and update manually this time.

**Comment by henrypp** [comment]
@webinger here is topic #644

**Comment by Kein** [comment]
> Any kind of software can be changed WF filter configuration, not even viruses or any kind of harmfull software, it can be done by any software you are installed (steam, utorrent, [even nsis installer have feature like this](https://nsis.sourceforge.io/NsisFirewall_plug-in), etc).

That still requires elevated privilegies and UAC prompt, no?
